File: demo/csharp-library/src/after/Validation.cs
// Validation utilities - pure functions
using System;

namespace UserManagement.Validation
{
    public static class UserValidator
    {
        public static void ValidateId(string id)
        {
            if (string.IsNullOrEmpty(id))
                throw new ArgumentException("ID is required");
        }

        public static void ValidateName(string name)
        {
            if (string.IsNullOrEmpty(name))
                throw new ArgumentException("Name is required");
        }

        public static void ValidateEmail(string email)
        {
            if (string.IsNullOrEmpty(email))
                throw new ArgumentException("Email is required");
            if (!email.Contains("@"))
                throw new ArgumentException("Invalid email format");
        }

        public static void ValidateAge(int age)
        {
            if (age <= 0 || age >= 150)
                throw new ArgumentException("Age must be between 1 and 149");
        }

        public static void ValidateRole(string role)
        {
            if (string.IsNullOrEmpty(role))
                throw new ArgumentException("Role is required");
            
            var validRoles = new[] { "Admin", "User", "Moderator" };
            if (Array.IndexOf(validRoles, role) == -1)
                throw new ArgumentException($"Invalid role: {role}");
        }

        public static void ValidateUser(User user)
        {
            if (user == null)
                throw new ArgumentNullException(nameof(user));
            
            ValidateId(user.Id);
            ValidateName(user.Name);
            ValidateEmail(user.Email);
            ValidateAge(user.Age);
            ValidateRole(user.Role);
        }
    }
}

File: demo/typescript-service/src/after/validation.ts
// Validation utilities - pure functions
export function validateUserId(id: string | undefined): void {
  if (!id) throw new Error('ID is required');
}

export function validateUserName(name: string | undefined): void {
  if (!name) throw new Error('Name is required');
}

export function validateUserEmail(email: string | undefined): void {
  if (!email) throw new Error('Email is required');
  if (!email.includes('@')) throw new Error('Invalid email format');
}

export function validateUserAge(age: number | undefined): void {
  if (!age) throw new Error('Age is required');
  if (age <= 0 || age >= 150) throw new Error('Invalid age');
}

export function validateUserRole(role: string | undefined): void {
  if (!role) throw new Error('Role is required');
  const validRoles = ['admin', 'user', 'moderator'];
  if (!validRoles.includes(role)) throw new Error('Invalid role');
}

export function validateUserData(userData: any): void {
  validateUserId(userData?.id);
  validateUserName(userData?.name);
  validateUserEmail(userData?.email);
  validateUserAge(userData?.age);
  validateUserRole(userData?.role);
}
